core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -208,10 +208,17 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
       if (!logInfos.isEmpty) {
         val newApps = new mutable.LinkedHashMap[String, FsApplicationHistoryInfo]()
         def addIfAbsent(info: FsApplicationHistoryInfo) = {
-          if (!newApps.contains(info.id) ||
-              newApps(info.id).logPath.endsWith(EventLoggingListener.IN_PROGRESS) &&
+          val key = 
+            if (info.appAttemptId.equals("")) { 
+              info.id 
+            } else {
+              info.id + "_" + info.appAttemptId
+            }
+          
+          if (!newApps.contains(key) ||
```

Comment:
Feels like you should add parentheses here somewhere to clarify in which order the conditions should be parsed, since you're mixing `$$` and `||`.